### PR TITLE
Cpd documentation

### PR DESCRIPTION
**Pull request by Rogal27**

### CODE REVIEW

File: doc/documentation.tex [hunk 1]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
 
 \subsection{Noise and outliers tolerance}
+We tested our methods using the clouds and conditions that may be closer to reality. That is, we used clouds that are missing points or even entire part and we corrupted clouds by noise and outliers. For this purpose we have used CPU versions of algorithms. For ICP and NICP \textit{policy} was set to \textit{parallel} and \textit{approximation-type} to \textit{none}. For CPD \textit{approximation-type} was set to \textit{hybrid}, mainly to achieve faster computation times. Many tests were conducted and it turned out that results are highly dependent on configuration. Therefore, we will show the most interesting results and tests that have shown the most consistent characteristics. All results are available in spreadsheet kept on Github: \href{https://github.com/Sliwson/cuda-slam/tree/cpd-documentation/doc/noise}{https://github.com/Sliwson/cuda-slam/tree/cpd-documentation/doc/noise}. Also in directory \textit{config} there are all test configurations.\\
```
Comment (szymon159) [inline doc/documentation.tex:420]: - points or even entire part -> points, even large part of them, 
- we will show you the -> we present you with the
- last sentence maybe like "All the test configurations are to be found in directory \textit{config} of the github repository".

File: doc/documentation.tex [hunk 2]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+After testing we came to several conclusions. Firstly, the ICP algorithm coped the worst with almost all test sets. The main reason can be cloud sizes used in tests, which occurred to be often too big for ICP capabilities. Algorithm had also trouble with noise. Even low level noise such as \textit{noise-affected-points-before}$=0.1$ and \textit{noise-intensity-before}$=0.05$ may cause convergence failure. Secondly, increasing $max-iterasions$ can help achieving more exact results. CPD algorithm mostly achieved correct results. Main key to CPD convergence is to select parameters value correctly. The most important one is \textit{cpd-weight} which describes how many points
```
Comment (szymon159) [inline doc/documentation.tex:421]: - Maybe: coped the worst with -> gave the worst results for
- low level noise ->  low noise level
- max-iterasions -> max-iterations
- start the part about CPD from new paragraph, fix the broken line 421-422

File: doc/documentation.tex [hunk 3]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+are noise and outliers. Difference of $0.3$ can lead to significantly different result. Second CPD parameter is \textit{cpd-const-scale}. Setting it to \textit{true} may help achieving better results when we know that the scale did not change after transformation. Especially it helps when clouds contain a lot of noise and outliers. Also, decresing \textit{cpd-tolerance} and \textit{converge-epsilon} (the latter applies also to ICP and NICP) may result in increasing iterations count and lead to more exact result.\\
+All algorithms had tendency to match areas where the point density was the highest. An example of this behavior may be observed using \textit{Config21.json}. It uses two "airbus" clouds with decreased sizes and added noise and outliers. Result achieved by algorithms are presented below and vary among each other.
```
Comment (szymon159) [inline doc/documentation.tex:423]: - point density -> points density
- do we name config files using upper case? maybe we should go for lower case if it is not a lot of work

File: doc/documentation.tex [hunk 4]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+	\centering
+    \subfloat[\centering CPD]{{\includegraphics[width=5cm]{mr-cpd-21.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering ICP]{{\includegraphics[width=5cm]{mr-icp-21.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering NICP]{{\includegraphics[width=5cm]{mr-nicp-21.png} }}
+    \caption{Tests with \textit{Config21.json}}
+\end{figure}
+
+The ICP did not manage to converge in this example but result achieved by CPD is interesting. Algorithm connected wheels and engines of both airplanes. Feasibly that is because those are areas with highest point density in entire cloud. NICP returned exact transformation.
```
Comment (szymon159) [inline doc/documentation.tex:435]: - the highest
- the entire cloud

File: doc/documentation.tex [hunk 5]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+
+Similar situation was met using \textit{Config11.json} and \textit{Config18.json}. Unfortunately in tests using the latter ICP returned identity matrix. The results are presented below.
```
Comment (szymon159) [inline doc/documentation.tex:437]: - the latter -> the latter one,

File: doc/documentation.tex [hunk 6]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+
+\begin{figure}[H]
+	\centering
+    \subfloat[\centering CPD]{{\includegraphics[width=6cm]{mr-cpd-18.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering NICP]{{\includegraphics[width=6cm]{mr-nicp-18.png} }}
+    \caption{Tests with \textit{Config18.json}}
+\end{figure}
+
+In contrary to \textit{Config18.json} and \textit{Config21.json}, using \textit{Config11.json} CPD almost found exact solution while this time NICP focused on matching areas with highest density.
```
Comment (szymon159) [inline doc/documentation.tex:457]: - almost found -> found almost

File: doc/documentation.tex [hunk 7]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+	\centering
+    \subfloat[\centering CPD]{{\includegraphics[width=5cm]{mr-cpd-26.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering ICP]{{\includegraphics[width=5cm]{mr-icp-26.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering NICP]{{\includegraphics[width=5cm]{mr-nicp-26.png} }}
+    \caption{Tests with \textit{Config26.json}}
+\end{figure}
+
+CPD found exact solution. As mentioned above ICP has trouble with incomplete clouds and it remains unchanged with this configuration. NICP was close to finding exact solution; however, it lacks of $90\degree$ rotation.
```
Comment (szymon159) [inline doc/documentation.tex:471]: - the exact solution
- ',' after mentioned above
- trouble -> troubles

File: doc/documentation.tex [hunk 8]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+
+Finally we made test that only focus on situation where one cloud contains noise. We increased the \textit{noise-intensity} by small values and compared the results. The best method occurred to be NICP, CPD was the second and ICP was the third one. For tests, the configurations \textit{Config28.json} to \textit{Config38} were used. Results are presented below.
```
Comment (szymon159) [inline doc/documentation.tex:473]: - for tests -> for these tests

File: doc/documentation.tex [hunk 9]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+	\centering
+    \subfloat[\centering CPD]{{\includegraphics[width=5cm]{mr-cpd-38.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering ICP]{{\includegraphics[width=5cm]{mr-icp-38.png} }}
+    \hspace{0.5em}
+    \subfloat[\centering NICP]{{\includegraphics[width=5cm]{mr-nicp-38.png} }}
+    \caption{Tests with \textit{Config38.json}}
+\end{figure}
+
+NICP managed to find exact solution in all test cases. Results obtained by CPD were slightly worse and sometimes correction of \textit{cpd-weight} parameter was needed. ICP results started to differ even when \textit{noise-intensity} was still very small, thus its usage when conditions are not perfect is very limited.
```
Comment (szymon159) [inline doc/documentation.tex:515]: - the exact solution

File: doc/documentation.tex [hunk 10]
```diff
@@ -466,6 +417,105 @@ \subsubsection{Methods comparison}
...
+
+To sum up the best algorithms for clouds with noises and outliers are NICP and CPD. It is worth mentioning that NICP is significantly faster than CPD.
```
Comment (szymon159) [inline doc/documentation.tex:517]: Hmm, the last sentence somehow doesn't fit. Maybe something like: "In this situation, it is also worth mentioning that NICP was significantly faster then CPD in all the tests covered in this section". Something like this, not sure how to do it exactly but this sentence looks just like I've added it here to flex with NICP being the fastest 😎

**Review by szymon159 — CHANGES_REQUESTED** [review]
Merge fixes have to be done as we cannot lose the changes in other sections.
The documentation part is good however, mainly small grammar improvements can be done but generally okay.

File: doc/documentation.tex [hunk 11]
```diff
@@ -192,18 +193,20 @@ \subsection{Non-iterative closest point}
...
-\begin{enumerate}
-\item \textbf{Full} - to avoid finding corresponding points in each algorithm run, this approximation type bases the error calculation on distance between two points with the same indexes in both clouds. It offers promising results with fast calculations for insufficiently permuted clouds but otherwise the error approximation might differ from actual one and for this reason returned result might not be the best.
-\item \textbf{None} - after each repetition, error is calculated by finding correspondences between subcloud of given size taken from first cloud and the entire second cloud. This method offers higher reliability but instead increases the execution time of one iteration noticeably, especially using CPU.
-\item \textbf{Hybrid} - error for each run is calculated the same way as using when using full approximation but 5 best results are stored and the error is recalculated for those in the same as when using none as approximation type (with finding correspondences). Then, the one with the smallest error is returned as an actual result.
-\end{enumerate}
-
```
[the remark this answers is not in the document]
Comment (Rogal27) [reply doc/documentation.tex:201]: Done

File: doc/documentation.tex [hunk 12]
```diff
@@ -220,20 +223,8 @@ \section{Execution, configuration and user guide}
...
-\includegraphics[width=\textwidth]{ss-exec.png}
-\caption{Execution example with given config file}
-\end{figure}
-\end{itemize} 
-
-Support for Linux systems and other development tools are some of the areas of improvement for the future. Porting, however, is not supposed to be an issue. But due to multiple dependencies between used libraries we decided not to focus on it in development phase as it is not the point of this research.
-
 \subsection{Configuration}
-Possible fields and values for config files are defined by \textit{JSON Schema} and they are described in the table below.
+Configuration of our project is provided by \textit{JSON} files. A path to configuration can be passed by command argument. If path is not provided the default configuration is used. Possible fields and values are defined by \textit{JSON Schema} and they are described in the table below.
```
[the remark this answers is not in the document]
Comment (Rogal27) [reply doc/documentation.tex:227]: Done

File: doc/documentation.tex [hunk 13]
```diff
@@ -242,126 +233,86 @@ \subsection{Configuration}
...
- 4 & Toggle visibility of the (0,0) point. \\
- \hline
- [ & Reduce the point size. \\
- \hline
- ] & Increase the point size. \\
- \hline
- \caption{Renderer controls}
-\end{longtable}
-\label{tab:controls}
-\end{center}
```
[the remark this answers is not in the document]
Comment (Rogal27) [reply doc/documentation.tex:364]: Done

File: doc/documentation.tex (hunk 1 above)
Comment (Rogal27) [reply doc/documentation.tex:420]: Done

File: doc/documentation.tex (hunk 2 above)
Comment (Rogal27) [reply doc/documentation.tex:421]: Done

File: doc/documentation.tex (hunk 3 above)
Comment (Rogal27) [reply doc/documentation.tex:423]: Done

File: doc/documentation.tex (hunk 4 above)
Comment (Rogal27) [reply doc/documentation.tex:435]: Done

File: doc/documentation.tex (hunk 5 above)
Comment (Rogal27) [reply doc/documentation.tex:437]: Done

File: doc/documentation.tex (hunk 6 above)
Comment (Rogal27) [reply doc/documentation.tex:457]: Done

File: doc/documentation.tex (hunk 7 above)
Comment (Rogal27) [reply doc/documentation.tex:471]: Done

File: doc/documentation.tex (hunk 8 above)
Comment (Rogal27) [reply doc/documentation.tex:473]: Done

File: doc/documentation.tex (hunk 9 above)
Comment (Rogal27) [reply doc/documentation.tex:515]: Done

File: doc/documentation.tex (hunk 10 above)
Comment (Rogal27) [reply doc/documentation.tex:517]: Done

File: doc/documentation.tex [hunk 14]
```diff
@@ -417,10 +476,10 @@ \subsubsection{Methods comparison}
...
+We tested our methods using the clouds and conditions that may be closer to reality. That is, we used clouds that are missing points, even large part of them and we corrupted clouds by noise and outliers. For this purpose we have used CPU versions of algorithms. For ICP and NICP \textit{policy} was set to \textit{parallel} and \textit{approximation-type} to \textit{none}. For CPD \textit{approximation-type} was set to \textit{hybrid}, mainly to achieve faster computation times. Many tests were conducted and it turned out that results are highly dependent on configuration. Therefore, in the following sections you can find the most interesting results and tests that have shown the most consistent characteristics. All results are available in spreadsheet kept on Github: \href{https://github.com/Sliwson/cuda-slam/tree/cpd-documentation/doc/noise}{https://github.com/Sliwson/cuda-slam/tree/cpd-documentation/doc/noise}. All the test configurations are to be found in directory \textit{config} of the Github repository. \\
```
Comment (szymon159) [inline doc/documentation.tex:479]: Missing coma after "large part of them"

File: doc/documentation.tex [hunk 15]
```diff
@@ -417,10 +476,10 @@ \subsubsection{Methods comparison}
...
+After testing we came to several conclusions. Firstly, the ICP algorithm gave the worst result for almost all test sets. The main reason can be cloud sizes used in tests, which occurred to be often too big for ICP capabilities. Algorithm had also trouble with noise. Even low noise level such as \textit{noise-affected-points-before}$=0.1$ and \textit{noise-intensity-before}$=0.05$ may cause convergence failure. Secondly, increasing \textit{max-iterations} can help achieving more exact results.\\
```
Comment (szymon159) [inline doc/documentation.tex:480]: - worst resultS

File: doc/documentation.tex (hunk 14 above)
Comment (Rogal27) [reply doc/documentation.tex:479]: Done

**Review by szymon159 — APPROVED** [review]

**Pull request merged**